dynamic_view() looses constness of its argument, later leading to compilation failure

Here is a minimal example:
```c++
const xt::xarray<int> a{1,2,3,4};
xt::slice_vector sv;
sv.push_back(xt::range(0,4,2));
auto v = xt::dynamic_view(a, sv);
auto iter = v.begin();
```
The last line fails to compile in `xstrided_view::data_xbegin()` and `xstrided_view::data_xend()`.

There is a similar error in this code (not sure if it is related to the above):
```c++
xt::slice_vector sv;
sv.push_back(xt::range(_,_,2));
view_nd<int, 3> v = ...;
auto dv = xt::dynamic_view(v, sv);
auto i = dv.begin();    // ok
auto const & cv = dv;
auto ci = cv.begin();   // doesn't compile
```
The error message says:
```
xstrided_view.hpp:629:47: error: could not convert from 
xt::xiterator<xt::xindexed_stepper<xvigra::view_nd<int, 3>, false>, 
              xvigra::tiny_vector<long int, 3, void>*, (xt::layout_type)1> 
to 
xt::xiterator<xt::xindexed_stepper<xvigra::view_nd<int, 3>, true>, 
              xvigra::tiny_vector<long int, 3, void>*, (xt::layout_type)1>
```

--- a/include/xtensor/xstrided_view.hpp
+++ b/include/xtensor/xstrided_view.hpp
@@ -196,7 +196,9 @@ namespace xt
         std::enable_if_t<detail::is_indexed_stepper<STEP>::value, STEP>
         stepper_end(const ST& shape, layout_type l) const;
 
-        using container_iterator = typename std::decay_t<CD>::iterator;
+        using container_iterator = std::conditional_t<is_const,
+                                                      typename std::decay_t<CD>::const_iterator,
+                                                      typename std::decay_t<CD>::iterator>;
         using const_container_iterator = typename std::decay_t<CD>::const_iterator;
 
         underlying_container_type& data() noexcept;

--- a/test/test_xdynamic_view.cpp
+++ b/test/test_xdynamic_view.cpp
@@ -204,6 +204,9 @@ namespace xt
         arr2 = dynamic_view(arr, {0});
         EXPECT_EQ(ref, arr2);
         // check that the following compiles
+        auto v = dynamic_view(arr, {0});
+        double acc = v(0);
+        EXPECT_EQ(acc, 2.5);
         auto iter = v.begin();
     }
 

--- a/test/test_xdynamic_view.cpp
+++ b/test/test_xdynamic_view.cpp
@@ -203,6 +203,8 @@ namespace xt
         xtensor<double, 2> ref{{2, 3}, 2.5};
         arr2 = dynamic_view(arr, {0});
         EXPECT_EQ(ref, arr2);
+        // check that the following compiles
+        auto iter = v.begin();
     }
 
     TEST(xdynamic_view, newaxis)

--- a/test/test_xview.cpp
+++ b/test/test_xview.cpp
@@ -367,6 +367,8 @@ namespace xt
         auto a2 = xt::view(a1, 1, xt::range(1, 2));
         int v2 = a2(0);
         EXPECT_EQ(v2, 3);
+
+        auto it = a2.begin();
     }
 
     TEST(xview, trivial_iterating)

